pantheon-bundle/src/main/java/com/redhat/pantheon/servlet/module/VanityVariantJsonServlet.java

```diff
@@ -0,0 +1,143 @@
+import com.redhat.pantheon.model.api.SlingModels;
+import com.redhat.pantheon.model.module.ModuleVariant;
+import com.redhat.pantheon.model.module.ModuleVersion;
+import com.redhat.pantheon.servlet.MapConverters;
+import com.redhat.pantheon.servlet.util.SlingPathSuffix;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.api.servlets.SlingSafeMethodsServlet;
+import org.apache.sling.servlets.annotations.SlingServletPaths;
+import org.jetbrains.annotations.NotNull;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+
+import javax.jcr.RepositoryException;
+import javax.servlet.Servlet;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.redhat.pantheon.servlet.ServletUtils.writeAsJson;
+
+/**
+ * An API endpoint for finding module information using a vanity url format which takes into account the content's
+ * product metadata, and locale. The format is as follows:<br><br>
+ *
+ * '/api/module/vanity.json/{locale}/{product}/{version}/{vanityUrlFragment}'<br><br>
+ *
+ * It will also resolve content with the following format: <br><br>
+ *
+ * '/api/module/vanity.json/{locale}/{product}/{version}/{moduleUUID}'
+ *
+ * <br>
+ * @see VariantJsonServlet for a servlet returning the same information
+ * @author Carlos Munoz
+ */
+@Component(
+        service = Servlet.class,
+        property = {
+                Constants.SERVICE_DESCRIPTION + "=Servlet to facilitate GET operation which accepts several path parameters to fetch module variant data",
+                Constants.SERVICE_VENDOR + "=Red Hat Content Tooling team"
+        })
+@SlingServletPaths(value = "/api/module/vanity")
+// /api/module/vanity.json/{locale}/{productLabel}/{versionLabel}/{vanityUrl}
+public class VanityVariantJsonServlet extends SlingSafeMethodsServlet {
+
+    private final SlingPathSuffix suffix = new SlingPathSuffix("/{locale}/{productLabel}/{versionLabel}/{vanityUrl}");
+
+    @Override
+    protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) throws ServletException, IOException {
+        Map<String, String> params = suffix.getParameters(request);
+        String locale = params.get("locale");
+        String productLabel = params.get("productLabel");
+        String versionLabel = params.get("versionLabel");
+        String vanityUrl = params.get("vanityUrl");
+
+        JcrQueryHelper queryHelper = new JcrQueryHelper(request.getResourceResolver());
+        try {
+
+            // Find the metadata with the right vanity url,
+            Optional<ModuleVersion> moduleVersion = queryHelper.query(
+                    "select * from [pant:moduleVersion] as v where v.[metadata/urlFragment] = '/" + vanityUrl + "'")
```

Comment:
A couple thoughts...
- Why are you using Module classes instead of Document classes (then it would hopefully work for Modules and Assemblies)?
- I see you're querying for the vanity url at the location we keep it today, which is on the metadata node. That does work, but I think I'd like to see a system where we manage vanity URLs (and redirects, etc) as separate entities in the system (perhaps located at /content/redirects in the JCR, something along those lines?) that contain pointers to the UUIDs of the DocumentVariants (would need to nail down what makes the most sense there) that they redirect to. In that world, this query would obviously need to be refactored. Regardless, the overall concept here is still solid.